feat: add accessor protocol support to `stats/base/varianceyc`

pulkitgupta2 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #5692 .

## Description

> What is the purpose of this pull request?

This pull request:

-   Refactors and adds accessor protocol support to `@stdlib/stats/base/varianceyc`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5692 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Upon reviewing the issue, I noted it references refactoring `@stats/base/varianceyv`, but this package doesn't exist in the repository. I've assumed this was a typographical error, and the intended target is `@stats/base/varianceyc` based on the context and references in the issue. I've proceeded with implementing the changes for `@stats/base/varianceyc` accordingly.


## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

pulkitgupta2 committed:
feat: add protocol support to stats/base/varianceyc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

pulkitgupta2 committed:
fixup! feat: add protocol support to stats/base/varianceyc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pulkitgupta2 committed:
fixup! fixup! feat: add protocol support to stats/base/varianceyc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pulkitgupta2 committed:
fixup! fixup! fixup! feat: add protocol support to stats/base/varianceyc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pulkitgupta2 committed:
fixup! fixup! fixup! fixup! feat: add protocol support to stats/base/varianceyc

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

jsndz commented:
@aayush0325  For review

aayush0325 reviewed (approved):
LGTM, thanks!

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/README.md:
```diff
@@ -114,17 +114,16 @@ The function has the following parameters:
+-   **strideX**: stride length for `x`.
 
-The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [variance][variance] of every other element in `x`,
+The `N` and `stride` parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [variance][variance] of every other element in `x`,
```
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [variance][variance] of every other element in `x`,
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/README.md:
````diff
@@ -114,17 +114,16 @@ The function has the following parameters:
+The `N` and `stride` parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [variance][variance] of every other element in `x`,
 
 ```javascript
 var floor = require( '@stdlib/math/base/special/floor' );
 
````
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/README.md:
```diff
@@ -139,13 +138,11 @@ var floor = require( '@stdlib/math/base/special/floor' );
 var x0 = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
```
The `floor` import above needs to be removed, as `floor` is no longer used.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/docs/types/index.d.ts:
```diff
@@ -20,7 +20,14 @@
-import { NumericArray } from '@stdlib/types/array';
+import { NumericArray, Collection, AccessorArrayLike } from '@stdlib/types/array';
+
+
```
```suggestion
```

We don't need two blank lines here.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/docs/types/index.d.ts:
```diff
@@ -20,7 +20,14 @@
+* Input array.
+*/
+type InputArray = NumericArray | Collection<number> | AccessorArrayLike<number>;
+
```
```suggestion
```

Nor here.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/docs/types/index.d.ts:
```diff
@@ -32,7 +39,7 @@ interface Routine {
 	* @param correction - degrees of freedom adjustment
 	* @param x - input array
-	* @param stride - stride length
+	* @param strideX - stride length for `x`
```
```suggestion
	* @param strideX - stride length
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/docs/types/index.d.ts:
```diff
@@ -41,16 +48,16 @@ interface Routine {
 	* @param x - input array
-	* @param stride - stride length
-	* @param offset - starting index
+	* @param strideX - stride length for `x`
```
```suggestion
	* @param strideX - stride length
```

kgryte replied:
Be consistent in descriptions. If you are going to use `for x`, then that needs to apply to both strides and offsets and all other applicable descriptions.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/README.md:
```diff
@@ -195,18 +190,13 @@ var v = varianceyc.ndarray( N, 1, x, 2, 1 );
-    x[ i ] = round( (randu()*100.0) - 50.0 );
-}
+var x = uniform( 10, -50.0, 50.0, {
+    'dtype': 'float64'
```
```suggestion
    'dtype': 'generic'
```

The original code generated a "generic" array.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/examples/index.js:
```diff
@@ -18,18 +18,12 @@
-	x[ i ] = round( (randu()*100.0) - 50.0 );
-}
+var x = uniform( 10, -50.0, 50.0, {
+	'dtype': 'float64'
```
```suggestion
	'dtype': 'generic'
```

The original code generated a "generic" array.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/lib/accessors.js:
```diff
@@ -0,0 +1,94 @@
+	var n;
+	var i;
+
+	n = N - correction;
+	if ( N <= 0 || n <= 0.0 ) {
+		return NaN;
+	}
+	if ( N === 1 || strideX === 0 ) {
+		return 0.0;
+	}
+
```
This logic isn't necessary, as we can centralize it in `ndarray.js`.
```suggestion
	n = N - correction;

```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/lib/ndarray.js:
```diff
@@ -47,31 +53,38 @@
 	var i;
+	var o;
+
+	o = arraylike2object( x );
+	if ( o.accessorProtocol ) {
+		return accessors( N, correction, o, strideX, offsetX );
+	}
 
```
This logic should be moved to L78 so that we don't need to repeat it in `accessors.js`.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -92,6 +130,25 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+	var x;
+	var v;
+
+	x = toAccessorArray([ 1.0, -2.0, -4.0, 5.0, 3.0 ]);
```
```suggestion
	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -92,6 +130,25 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+	v = varianceyc( -1, 1, x, 1, 0 );
+	t.strictEqual( isnan( v ), true, 'returns expected value' );
+
+	x = toAccessorArray([ 10.0 ]);
```
```suggestion
	x = toAccessorArray( [ 10.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -104,6 +161,22 @@ tape( 'if provided an `N` parameter equal to `1`, the function returns a populat
+	var x;
+	var v;
+
+	x = toAccessorArray([ 1.0, -2.0, -4.0, 5.0, 3.0 ]);
```
```suggestion
	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
```

This same change applies throughout this file. Use consistent spacing.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -92,6 +130,25 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+
+	x = toAccessorArray([ 10.0 ]);
+	v = varianceyc( 0, 0, x, 1, 0 );
+	t.strictEqual( isnan( v ), true, 'returns expected value when N = 0 with a single value' );
```
```suggestion
	t.strictEqual( isnan( v ), true, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -104,6 +161,22 @@ tape( 'if provided an `N` parameter equal to `1`, the function returns a populat
+
+	x = toAccessorArray([ 5.0 ]);
+	v = varianceyc( 1, 0, x, 1, 0 );
+	t.strictEqual( v, 0.0, 'returns expected value with a single value' );
```
```suggestion
	t.strictEqual( v, 0.0, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -119,8 +192,26 @@ tape( 'if provided a `correction` parameter yielding `N-correction` less than or
+	x = toAccessorArray([ 1.0, -2.0, -4.0, 5.0, 3.0 ]);
+
+	v = varianceyc( x.length, x.length, x, 1, 0 );
+	t.strictEqual( isnan( v ), true, 'returns expected value when n = 0' );
```
```suggestion
	t.strictEqual( isnan( v ), true, 'returns expected value' );
```

No bespoke messages unless actually necessary. Applies throughout this PR.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js:
```diff
@@ -192,10 +333,50 @@ tape( 'the function supports an `offset` parameter', function test( t ) {
+	t.end();
+});
+
+tape( 'the function handles specific accessor cases for coverage', function test( t ) {
```
This can be removed after you refactor the implementation to avoid duplicated logic.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.varianceyc.js:
```diff
@@ -93,6 +131,25 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+	var x;
+	var v;
+
+	x = toAccessorArray([ 1.0, -2.0, -4.0, 5.0, 3.0 ]);
```
```suggestion
	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
```

Same comments here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.varianceyc.js:
```diff
@@ -93,6 +131,25 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+
+	x = toAccessorArray([ 10.0 ]);
+	v = varianceyc( 0, 0, x, 1 );
+	t.strictEqual( isnan( v ), true, 'returns expected value when N = 0 with a single value' );
```
```suggestion
	t.strictEqual( isnan( v ), true, 'returns expected value' );
```

Same comments here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/stats/base/varianceyc/test/test.varianceyc.js:
```diff
@@ -197,10 +338,57 @@ tape( 'the function supports view offsets', function test( t ) {
 	t.end();
 });
+
+tape( 'the function handles specific accessor cases for complete branch coverage', function test( t ) {
```
Same comments. Can be removed after refactoring.

kgryte reviewed (changes requested):
@pulkitgupta2 Thank you for opening this PR; however, this needs a fair amount of clean-up before we can merge. Once all the comments are resolved, we can perform another review.

cc @aayush0325

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/README.md

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/README.md

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/docs/types/index.d.ts

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/README.md

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/examples/index.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
Update lib/node_modules/@stdlib/stats/base/varianceyc/test/test.ndarray.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
fixup! refactor: updated docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pulkitgupta2 committed:
fixup! refactor: updated accessor implemntation to avoid duplication

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

pulkitgupta2 commented:
Hi @kgryte. The comments added are resolved. 

CC @aayush0325

pulkitgupta2 committed:
Merge branch 'develop' into refactor-varianceyc

Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
refac: formatting changes

Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

pulkitgupta2 committed:
refac: formatting changes

Signed-off-by: Pulkit Gupta <65711278+pulkitgupta2@users.noreply.github.com>

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (dismissed):
LGTM after clean-up.

kgryte commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years
